src/ethereum/frontier/eth_types.py

```diff
@@ -120,7 +120,24 @@ class Receipt:
     post_state: Root
     cumulative_gas_used: Uint
     bloom: Bloom
-    logs: List[Log]
+    logs: Tuple[Log, ...]
 
 
 State = Dict[Address, Account]
+
+
+def increase_balance(state: State, address: Address, amount: U256) -> None:
+    """Increase the balance of an account in the State."""
+    account = state[address]
+    state[address] = Account(
+        account.nonce, account.balance + amount, account.code, account.storage
+    )
+
+
+def reduce_balance(state: State, address: Address, amount: U256) -> None:
+    """Reduce the balance of an account in the State."""
+    account = state[address]
+    assert account.balance >= amount
+    state[address] = Account(
+        account.nonce, account.balance - amount, account.code, account.storage
+    )
```

Comment:
It'd be even better to include the actual state update in the transformation as well, so maybe this is better suited as a method on `State`, and we ignore my second suggestion?

Something like:

```python
@dataclass                               # Don't freeze this class.
class MutableAccount(Account):           # Might need to reverse the inheritance order, I'm not sure.
    pass

def state_update(state: State, address: Address, transform: Callable[[MutableAccount], None]):
    # ...
```

Reply:
There are basically two ways to do this

## Freeze/thaw

There is a function `thaw()` that takes a immutable object and makes a mutable copy and a function `freeze()` which makes an object immutable (without copying). You can then write the following:

```python
account = thaw(state[address])
account.balance += BLOCK_REWARD
account.freeze()
state[address] = account
```
## Replace/update

The state object comes with two functions, `replace()` and `update()`. `replace()` works like `dataclasses.replace()`. `update()` takes mutator functions rather than new values. For example:

```python
state.replace(address, balance=state[address].balance + BLOCK_REWARD)
```

```python
state.update(address, balance=lambda x: x + BLOCK_REWARD)
```

## Thoughts

I prefer the replace/update method to freeze/thaw. It's neater and avoids the possibility of the following bug:

```python
account = thaw(state[address])
account.balance += BLOCK_REWARD

do_something_to_account(address)

account.freeze()
state[address] = account
```
The effects of `do_something_to_account()` get clobbered.